dsgrid/spark/functions.py

```diff
@@ -0,0 +1,532 @@
+        #        df = df.withColumn(field.name, F.col(field.name).cast(TimestampType()))
+        df = spark.createDataFrame(pd.read_csv(path_str))
+        if "timestamp" in df.columns:
+            # TODO duckdb: do something better
+            df = df.withColumn("timestamp", F.col("timestamp").cast(TimestampType()))
+        dup_cols = [x for x in df.columns if x.endswith(".1")]
+        if dup_cols:
+            msg = f"Detected a duplicate column in the dataset: {dup_cols}"
+            raise DSGInvalidDimension(msg)
+        return df
+    return spark.read.csv(str(path), header=True, inferSchema=True)
+
+
+def read_json(path: Path | str) -> DataFrame:
+    """Return a DataFrame from a JSON file, handling special cases with duckdb.
+
+    Warning: Use of this function with DuckDB is not efficient because it requires that we
+    convert line-delimited JSON to standard JSON.
+    """
+    spark = get_spark_session()
+    filename = str(path)
+    if use_duckdb():
+        with NamedTemporaryFile(suffix=".json") as f:
+            f.close()
+            data = load_line_delimited_json(path)
+            dump_data(data, f.name)
+            return spark.read.json(f.name)
+    return spark.read.json(filename, mode="FAILFAST")
+
+
+def read_parquet(path: Path | str) -> DataFrame:
+    path = Path(path) if isinstance(path, str) else path
+    spark = get_spark_session()
+    if path.is_file() or not use_duckdb():
+        df = spark.read.parquet(str(path))
+    else:
+        df = spark.read.parquet(f"{path}/**/*.parquet")
+    return df
+
+
+def select_expr(df: DataFrame, exprs: list[str]) -> DataFrame:
+    if use_duckdb():
+        view = create_temp_view(df)
+        spark = get_spark_session()
+        cols = ",".join(exprs)
+        return spark.sql(f"SELECT {cols} FROM {view}")
+    return df.selectExpr(*exprs)
+
+
+def sql_from_df(df: DataFrame, query: str) -> DataFrame:
+    """Run a SQL query on a dataframe with Spark."""
+    logger.debug("Run SQL query [%s]", query)
+    spark = get_spark_session()
+    if use_duckdb():
+        view = create_temp_view(df)
+        query += f" FROM {view}"
+        return spark.sql(query)
+
+    query += " FROM {df}"
+    return spark.sql(query, df=df)
+
+
+def pivot(df: DataFrame, name_column: str, value_column: str) -> DataFrame:
+    """Unpivot the dataframe."""
+    method = _pivot_duckdb if use_duckdb() else _pivot_spark
+    return method(df, name_column, value_column)
+
+
+def _pivot_duckdb(df: DataFrame, name_column: str, value_column: str) -> DataFrame:
+    view = create_temp_view(df)
+    query = f"""
+        PIVOT {view}
+        ON {name_column}
+        USING SUM({value_column})
+    """
+    return get_spark_session().sql(query)
+
+
+def _pivot_spark(df: DataFrame, name_column: str, value_column: str) -> DataFrame:
+    ids = [x for x in df.columns if x not in {name_column, value_column}]
+    return df.groupBy(*ids).pivot(name_column).sum(value_column)
```

Comment:
This would not require an aggregation, 
`df.pivot(index=*ids, columns=name_column, values="value_column)`

https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/api/pyspark.pandas.DataFrame.pivot.html?highlight=pivot#pyspark.pandas.DataFrame.pivot

Reply:
Your example is from their pandas API, but I'll look into it.